contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,384 @@
+        path[0] = WETH;
+        path[1] = _token;
+        uint256 amountOut = _router.swapExactTokensForTokens(_amount, 0, path, address(this), block.timestamp)[1];
+        return amountOut;
+    }
+ 
+    function purchaseTokenBestPrice(address _token, uint256 _amount) internal returns (uint256) {
+        uint256 uniPrice = tokenAvailable(uniFactory, _token) ? getBuyPrice(true, _token, _amount) : PreciseUnitMath.maxUint256();
+        uint256 sushiPrice = tokenAvailable(sushiFactory, _token) ? getBuyPrice(false, _token, _amount) : PreciseUnitMath.maxUint256();
+        if (uniPrice <= sushiPrice) {
+            return purchaseToken(uniRouter, _token, _amount);
+        } else {
+            return purchaseToken(sushiRouter, _token, _amount);
+        }
+    }
+
+    function sellToken(IUniswapV2Router02 _router, address _token) internal {
+        uint256 tokenBalance = IERC20(_token).balanceOf(address(this));
+        address[] memory path = new address[](2);
+        path[0] = _token;
+        path[1] = WETH;
+        _router.swapExactTokensForTokens(tokenBalance, 0, path, address(this), block.timestamp);
+    }
+
+    function sellTokenBestPrice(address _token) internal {
+        uint256 tokenBalance = IERC20(_token).balanceOf(address(this));
+        uint256 uniPrice = tokenAvailable(uniFactory, _token) ? getSellPrice(true, _token, tokenBalance) : 0;
+        uint256 sushiPrice = tokenAvailable(sushiFactory, _token) ? getSellPrice(false, _token, tokenBalance) : 0;
+        if (uniPrice >= sushiPrice) {
+            sellToken(uniRouter, _token);
+        } else {
+            sellToken(sushiRouter, _token);
+        }
+    }
+
+    /**
+     * Purchases an Ether given an exact amount of a token to spend
+     *
+     * @param _token    token to spend
+     * @param _amount   amount of token to spend
+     */
+    function purchaseWETHExactTokens(address _token, uint256 _amount) internal {
+        address[] memory path = new address[](2);
+        path[0] = _token;
+        path[1] = WETH;
+
+        uint256 uniAmountOut = tokenAvailable(uniFactory, _token) ? uniRouter.getAmountsOut(_amount, path)[1] : 0;
+        uint256 sushiAmountOut = tokenAvailable(sushiFactory, _token) ? sushiRouter.getAmountsOut(_amount, path)[1] : 0;
+        IUniswapV2Router02 router = uniAmountOut >= sushiAmountOut ? uniRouter : sushiRouter;
+
+        IERC20(_token).approve(address(router), PreciseUnitMath.maxUint256());
+        router.swapExactTokensForTokens(_amount, 0, path, address(this), block.timestamp);
+    }
+
+    /**
+     * Gets the components of a set and the corresponding amount needed
+     *
+     * @param _setToken     set token
+     * @param _amount       amount of the set token
+     * @return              a tuple containing arrays of component token addresses and amounts of tokens needed
+     */
+    function getTokensNeeded(ISetToken _setToken, uint256 _amount) internal view returns (address[] memory, uint256[] memory) {
```

Comment:
To add precision to naming, consider renaming this to `getRequiredComponentUnitsForIssuance` like in the `BasicIssuanceModule` (link below)

https://github.com/SetProtocol/set-protocol-v2-legacy/blob/master/contracts/protocol/modules/BasicIssuanceModule.sol#L203